### PR TITLE
Network Simulator: class Event

### LINKED ISSUE #29: Network Simulator: create class Event

### PR DESCRIPTION
**Pull request by AntoxaBarin**
Closes #29. Implements events -- callbacks scheduled at some specified time in future.

### CODE REVIEW

File: include/Event.hpp [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+     */
+    std::uint64_t delivery_time_ns;
+
+    /**
+     * @brief Acknowledgement flag
+     *
+     * @details True: ACK
+     * @details False: Packet
+     */
+    bool is_ACK;
```
Comment (drgulevich) [inline include/Event.hpp:39]: I do not think `is_ack` is a property of an *event*. Should `is_ack` be a property of a *packet* instead?

**Review by drgulevich — COMMENTED** [review]
See https://github.com/Cloud-Storage-Team/algnet/blob/main/README.md

File: include/Event.hpp (hunk 1 above)
Comment (AntoxaBarin) [reply include/Event.hpp:39]: Yes, it should be a property of a `Packet`.

Comment (AntoxaBarin) [reply include/Event.hpp:39]: Yes, it should be a property of a `Packet`.

File: include/Event.hpp [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+#include <functional>
+
+class Event {
+public:
+    Event(std::uint64_t time, const std::function<void()>& handler);
+
+    std::uint64_t GetTime() const;
+    void Execute() const;
+
+    std::uint64_t execution_time;
```
Comment (drgulevich) [inline include/Event.hpp:16]: We either need a public getter `GetTime` and private `execution_time` OR public `execution_time`, but not both: if you use `GetTime` then  `execution_time` should be private.

File: include/Event.hpp [hunk 3]
```diff
@@ -0,0 +1,20 @@
+#pragma once
+
+#include "Packet.hpp"
+
+#include <memory>
```
Comment (drgulevich) [inline include/Event.hpp:5]: Which functions use this include file?

File: include/Event.hpp [hunk 4]
```diff
@@ -0,0 +1,20 @@
+#pragma once
+
+#include "Packet.hpp"
```
Comment (drgulevich) [inline include/Event.hpp:3]: Where do you use  "Packet.hpp"?

Comment (AntoxaBarin) [reply include/Event.hpp]: Now Event and Packet are independent classes, this include is left over from the previous Event implementation.

File: include/Event.hpp [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+#include <functional>
+
+class Event {
+public:
+    Event(std::uint64_t time, const std::function<void()>& handler);
+
+    void Execute() const;
+
+    std::uint64_t execution_time;
+    std::function<void()> handler;
```
Comment (drgulevich) [inline include/Event.hpp:13]: 1. Nit: m_handler
2. *m_handler* should be private

File: include/Event.hpp [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+#include <cstdint>
+#include <functional>
+
+class Event {
+public:
+    Event(std::uint64_t time, const std::function<void()>& handler);
+
+    void Execute() const;
+
+    std::uint64_t m_execution_time;
```
Comment (drgulevich) [inline include/Event.hpp:12]: Nit: `m_` for private members only to distinguish from public data (https://lefticus.gitbooks.io/cpp-best-practices/content/03-Style.html).

Comment (drgulevich) [inline include/Event.hpp:12]:
```suggestion
    std::uint64_t execution_time;
```

File: src/Event.cpp [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+Event::Event(std::uint64_t time, const std::function<void()> &handler):
+        m_execution_time(time),
+        m_handler(handler) { }
+
+void Event::Execute() const {
+    m_handler();
+}
+
+bool Event::operator<(const Event& other) const {
+    return m_execution_time > other.m_execution_time;
```
Comment (drgulevich) [inline src/Event.cpp:12]:
```suggestion
    return execution_time > other.execution_time;
```

**Review by drgulevich — APPROVED** [review]
Thanks

File: src/Event.cpp [hunk 8]
```diff
@@ -0,0 +1,13 @@
+#include "Event.hpp"
+
+Event::Event(std::uint64_t time, const std::function<void()> &handler):
+        m_execution_time(time),
```
Comment (drgulevich) [inline src/Event.cpp:4]:
```suggestion
        execution_time(time),
```

**Pull request merged**